README.md
# OctoPrint-LCDproc

## About

This plugin will enable support for the [LCDd/LCDproc](http://lcdproc.omnipotent.net) supported displays.

- [ChangeLog](CHANGELOG.md)
- [Issue Tracker](https://github.com/czo/octoprint-lcdproc/issues/)

## Screenshots

[<img src="blob/gLCD-without_title.png" width="250"/>](blob/gLCD-without_title.png)|[<img src="blob/gLCD-with_title.png" width="250"/>](blob/gLCD-with_title.png)
:--:|:--:
[<img src="blob/HW-24x2-LCD.jpeg" width="250"/>](blob/HW-24x2-LCD.jpeg)|[<img src="blob/settings_page.png" width="250"/>](blob/settings_page.png)

## Setup

Install manually using this URL ([Plugin Manager](https://docs.octoprint.org/en/master/bundledplugins/pluginmanager.html) will be available later):

    https://github.com/czo/octoprint-lcdproc/archive/refs/tags/0.1.3.zip

or for the current, development version:

    https://github.com/czo/octoprint-lcdproc/archive/refs/heads/main.zip
    

## Configuration

**TODO:** Describe your plugin's configuration options (if any).


CHANGELOG.md
# OctoPrint-LCDproc - Changelog

All major changes will be documented here.

## [0.1.4] - WP
- not yet known

## [0.1.3] - 2022-07-24
- Fixed a bug, which was related to multi-day prints. The displayed estimated finish time was a bit misleading when checked after midnight.

## [0.1.2] - 2022-07-19
- Python 2.7 compatibility fix 

## [0.1.1] - 2022-07-13
- moving code to github from my private gitlab
- fixed the completion progress display, it can go further than 99% to finally reach 100%

## [0.1.0] - 2022-07-13
- first public release
- working software update
- configuration page


octoprint_lcdproc/lcdproc/server.py
from __future__ import print_function
from urllib.parse import unquote
import select

from .screen import Screen
from .pytelnetlib import telnetlib

class Server(object):
    
    """ LCDproc Server Object """
    
    def __init__(self, hostname="localhost", port=13666, debug=False):
        
        """ Constructor """
        
        self.debug = debug
        self.hostname = hostname
        self.port = port
        self.tn = None
        self.server_info = dict()
        self.screens = dict()
        self.keys = list()
                
    def start_session(self):
        
        """ Start Session """

        self.tn = telnetlib.Telnet(self.hostname, self.port)
        
        response = self.request("hello") 
        bits = response.split(" ")
        self.server_info.update({
            "server_version": bits[2],
            "protocol_version": bits[4],
            "screen_width": int(bits[7]),
            "screen_height": int(bits[9]),
            "cell_width": int(bits[11]),
            "cell_height": int(bits[13])            
        })                
        return response  

    def close_session(self):
        if self.tn:
            self.tn.close()
        self.tn = None
        self.server_info = dict()
        self.screens = dict()
        self.keys = list()

    def alive_session(self):
        if not self.tn:
            return False
        if not self.server_info:
            return False
        return True

    def request(self, command_string):
        
        """ Request """
        if not self.tn:
            return
        try:
            self.tn.write((command_string + "\n").encode())
        except:
            self.close_session()
            return None

        if self.debug: print("Telnet Request:", command_string)
        while True:
            try:
                response = unquote(self.tn.read_until(b"\n").decode())
            except:
                self.close_session()
                return None
            if "success" in response:   # Normal successful reply
                break
            if "huh" in response:       # Something went wrong
                break
            if "connect" in response:   # Special reply to "hello"
                break
            # TODO Keep track of which screen is displayed
            # Try again if response was key, menu or visibility notification.
        if "huh" in response or self.debug: print("Telnet Response:", response[:-1])
        return response


    # def poll(self):
    #     """
    #     Poll
        
    #     Check for a non-response string generated by LCDd and return any string read.
    #     LCDd generates strings for key presses, menu events & screen visibility changes.
    #     """

    #     if not self.tn:
    #         return

    #     if select.select([self.tn], [], [], 0) == ([self.tn], [], []):
    #         response = unquote(self.tn.read_until(b"\n").decode())
    #         if self.debug: print("Telnet Poll:", response[:-1])
    #         # TODO Keep track of which screen is displayed
    #         return response
    #     else:
    #         return None

    def add_screen(self, ref):     

        if not self.tn:
            return
        
        """ Add Screen """
        
        if ref not in self.screens:   
            screen = Screen(self, ref)
            screen.clear()              # TODO Check this is needed, new screens should be clear.
            self.screens[ref] = screen
            return self.screens[ref]
     

    def del_screen(self, ref):
        """ Delete/Remove Screen """

        if not self.tn:
            return

        self.request("screen_del %s" % (ref))
        del(self.screens[ref])

        
    def add_key(self, ref, mode = "shared"):
        """
        Add a key.
        
        (ref)
        Return key name or None on error
        """

        if not self.tn:
            return

        if ref not in self.keys:   
            response = self.request("client_add_key -%s %s" % (mode, ref))
            if "success" not in response: return None
            self.keys.append(ref)
            return ref


    def del_key(self, ref):
        """
        Delete a key.
        
        (ref)
        Return None or LCDd response on error
        """

        if not self.tn:
            return

        if ref not in self.keys:   
            response = self.request("client_del_key %s" % (ref))
            self.keys.remove(ref)
            if "success" in response:
                return None
            else:
                return response
            

    def output(self, value):
        """
        Sets the client's output (on, off, int)
        
        Sets the general purpose output on some display modules to this value.
        Use on to set all outputs to high state, and off to set all to low state.
        The meaning of the integer value depends on your specific device, usually
        it is a bit pattern describing the state of each output line.
        
        Return None or LCDd response on error
        """

        if not self.tn:
            return

        response = self.request(("output %s" % (value)).encode())
        if "success" in response:
            return None
        else:
            return response

    def get_server_info(self):
        """
        Returns information about the server and the connected
        LCD as a dictionary with the following keys:

            "server_version", "protocol_version",
            "screen_width", "screen_height",
            "cell_width", "cell_height"
        """
        return self.server_info


octoprint_lcdproc/lcdproc/widgets.py
class StringWidget(object):

    """ String Widget """

    def __init__(self, screen, ref, x, y, text):
        
        self.screen = screen
        self.ref = ref
        self.x = x
        self.y = y
        self.text = text

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, self.ref, "string"))
        self.update()

    
    def update(self):
        self.screen.server.request('widget_set %s %s %s %s "%s"' % (self.screen.ref, self.ref, self.x, self.y, self.text))

    
    def set_x(self, x):
        self.x = x
                                   
    
    def set_y(self, y):
        self.y = y
        
    
    def set_text(self, text):
        self.text = text
        
        
class TitleWidget(object):
    
    """ Title Widget """
    
    def __init__(self, screen, ref, text):
        
        self.screen = screen
        self.ref = ref
        self.text = text

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, self.ref, "title"))
        self.update()
        
    def update(self):
        self.screen.server.request('widget_set %s %s "%s"' % (self.screen.ref, self.ref, self.text))

    def set_text(self, text):
        self.text = text
        
        
class HBarWidget(object):
    
    def __init__(self, screen, ref, x, y, length):
        
        self.screen = screen
        self.ref = ref
        self.x = x
        self.y = y
        self.length = length

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, self.ref, "hbar"))
        self.update()
        
    def update(self):
        
        self.screen.server.request("widget_set %s %s %s %s %s" % (self.screen.ref, self.ref, self.x, self.y, self.length))

    def set_x(self, x):
        
        self.x = x
                                   
    def set_y(self, y):
        
        self.y = y
        
    def set_length(self, length):
        
        self.length = length
        
                                      
class VBarWidget(object):
    
    def __init__(self, screen, ref, x, y, length):
        
        self.screen = screen
        self.ref = ref
        self.x = x
        self.y = y
        self.length = length

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, self.ref, "vbar"))
        self.update()
        
    def update(self):
        
        self.screen.server.request("widget_set %s %s %s %s %s" % (self.screen.ref, self.ref, self.x, self.y, self.length))

    def set_x(self, x):
        
        self.x = x
                                   
    def set_y(self, y):
        
        self.y = y
        
    def set_length(self, length):
        
        self.length = length
        
        
class IconWidget(object):
    
    def __init__(self, screen, ref, x, y, name):
        
        self.screen = screen
        self.ref = ref
        self.x = x
        self.y = y
        self.name = name

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, self.ref, "icon"))
        self.update()
        
    def update(self):
        
        self.screen.server.request("widget_set %s %s %s %s %s" % (self.screen.ref, self.ref, self.x, self.y, self.name))

    def set_x(self, x):
        
        self.x = x
                                   
    def set_y(self, y):
        
        self.y = y
        
    def set_name(self, name):
        
        self.name = name
        
class ScrollerWidget(object):
    
    def __init__(self, screen, ref, left, top, right, bottom, direction, speed, text):
        self.screen = screen
        self.ref = ref
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom
        self.direction = direction
        self.speed = speed
        self.text = text

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, 
                                                            self.ref, 
                                                            "scroller"))
        self.update()
        
    def update(self):
        self.screen.server.request('widget_set %s %s %s %s %s %s %s %s "%s"' % (self.screen.ref, 
                                                                  self.ref, 
                                                                  self.left, 
                                                                  self.top, 
                                                                  self.right, 
                                                                  self.bottom, 
                                                                  self.direction, 
                                                                  self.speed, 
                                                                  self.text))

    def set_left(self, left):
        self.left = left
                                   
    def set_top(self, top):
        self.top = top
        
    def set_right(self, right):
        self.right = right
        
    def set_bottom(self, bottom):
        self.bottom = bottom
        
    def set_direction(self, direction):
        self.direction = direction
        
    def set_speed(self, speed):
        self.speed = speed
        
    def set_text(self, text):
        self.text = text
        
        
class FrameWidget(object):
    
    def __init__(self, screen, ref, left, top, right, bottom, width, height, direction, speed):
        self.screen = screen
        self.ref = ref
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom
        self.width = width
        self.height = height
        self.direction = direction
        self.speed = speed

        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, 
                                                            self.ref, 
                                                            "frame"))
        self.update()
        
    def update(self):
        self.screen.server.request('widget_set %s %s %s %s %s %s %s %s %s %s' % (self.screen.ref, 
                                                                  self.ref, 
                                                                  self.left, 
                                                                  self.top, 
                                                                  self.right, 
                                                                  self.bottom, 
                                                                  self.width,
                                                                  self.height,
                                                                  self.direction, 
                                                                  self.speed))

    def set_left(self, left):
        self.left = left
                                   
    def set_top(self, top):
        self.top = top
        
    def set_right(self, right):
        self.right = right
        
    def set_bottom(self, bottom):
        self.bottom = bottom

    def set_width(self, width):
        self.width = width
        
    def set_height(self, height):
        self.height = height
        
    def set_direction(self, direction):
        self.direction = direction
        
    def set_speed(self, speed):
        self.speed = speed
                                          
                                          
class NumberWidget(object):
    
    def __init__(self, screen, ref, x, value):
        self.screen = screen
        self.ref = ref
        self.x = x
        self.value = value
        
        self.screen.server.request("widget_add %s %s %s" % (self.screen.ref, 
                                                            self.ref, 
                                                            "num"))
        self.update()
        
    def update(self):
        self.screen.server.request('widget_set %s %s %s %s' % (self.screen.ref, 
                                                               self.ref, 
                                                               self.x,
                                                               self.value))

    def set_x(self, x):
        self.x = x
                                   
    def set_value(self, value):
        self.value = value


octoprint_lcdproc/static/js/lcdproc.js
/*
 * View model for OctoPrint-LCDproc
 *
 * Author: Jozsef CZOMPO
 * License: GPLv3
 */
$(function() {
    function LcdprocViewModel(parameters) {
        var self = this;

        // assign the injected parameters, e.g.:
        // self.loginStateViewModel = parameters[0];
        // self.settingsViewModel = parameters[1];

        // TODO: Implement your plugin's view model here.
    }

    /* view model class, parameters for constructor, container to bind to
     * Please see http://docs.octoprint.org/en/master/plugins/viewmodels.html#registering-custom-viewmodels for more details
     * and a full list of the available options.
     */
    OCTOPRINT_VIEWMODELS.push({
        construct: LcdprocViewModel,
        // ViewModels your plugin depends on, e.g. loginStateViewModel, settingsViewModel, ...
        dependencies: [ /* "loginStateViewModel", "settingsViewModel" */ ],
        // Elements to bind to, e.g. #settings_plugin_lcdproc, #tab_plugin_lcdproc, ...
        elements: [ /* ... */ ]
    });
});


octoprint_lcdproc/__init__.py
# coding=utf-8
from __future__ import absolute_import
from datetime import datetime, timedelta

import octoprint.plugin
from octoprint.printer import PrinterInterface
from octoprint.events import Events
from octoprint.util import RepeatedTimer, ResettableTimer, get_formatted_datetime, get_formatted_timedelta

from octoprint_lcdproc.lcdproc.server import Server

STATE_NON_PRINTING = "non_printing"
STATE_PRINTING = "printing"
STATE_IDLE = "idle"

class LcdprocPlugin(octoprint.plugin.SettingsPlugin,
    octoprint.plugin.AssetPlugin,
    octoprint.plugin.TemplatePlugin,
    octoprint.plugin.StartupPlugin,
    octoprint.plugin.ShutdownPlugin,
    octoprint.plugin.EventHandlerPlugin,
    octoprint.plugin.ProgressPlugin
):

    lcd = None
    timer_screen = None
    timer_seconds = None
    screen_priority_state = STATE_IDLE
    start_timestamp = None
    printing_filename = None
    printing_percent = None
    printing_eta = None

    ##~~ SettingsPlugin mixin

    def get_settings_defaults(self):
        return {
            # put your plugin's default settings here
        }

    ##~~ AssetPlugin mixin

    def get_assets(self):
        # Define your plugin's asset files to automatically include in the
        # core UI here.
        return {
            "js": ["js/lcdproc.js"],
            "css": ["css/lcdproc.css"],
            "less": ["less/lcdproc.less"]
        }

    ##~~ Softwareupdate hook

    def get_update_information(self):
        # Define the configuration for your plugin to use with the Software Update
        # Plugin here. See https://docs.octoprint.org/en/master/bundledplugins/softwareupdate.html
        # for details.
        return {
            "lcdproc": {
                "displayName": self._plugin_name,
                "displayVersion": self._plugin_version,
                "current": self._plugin_version,

                "type": "jsondata",
                "jsondata": "https://raw.githubusercontent.com/czo/octoprint-lcdproc/main/softwareupdate.json",

                "method": "pip",
                "pip": "https://github.com/czo/octoprint-lcdproc/archive/refs/tags/{target}.zip",
                "release_notes": "https://github.com/czo/octoprint-lcdproc/blob/main/CHANGELOG.md",
            }
        }

    def get_settings_defaults(self):
        return {
            "enabled": True,
            "host": "localhost",
            "port": 13666,
            "hide_page_when_idle": True,
            "priority_printing": "foreground",
            "priority_non_printing": "info",
            "idle_time_minutes": 60,
            "title_show": False,
            "title_text": "OctoPrint",
        }

    def get_template_configs(self):
        return [
            {
                "type": "settings",
                "custom_bindings": False,
            }
        ]

    def on_settings_save(self, data):
        anything_changed = False
        current_values = self.get_settings_defaults()
        for key in self.get_settings_defaults().keys():
            current_values[key] = self._settings.get([key])
        octoprint.plugin.SettingsPlugin.on_settings_save(self, data)
        for key in self.get_settings_defaults().keys():
            if not current_values[key] == self._settings.get([key]):
                anything_changed = True
                break

        if anything_changed:
            self._logger.info("Configuration changed, destroying connection")
            if self.lcd:
                self.lcd.close_session()
                self.lcd = None

    def on_startup(self, host, port):
        self.initialize_lcd()

    def on_event(self, event, payload):
        if event in [ Events.PRINT_STARTED, Events.PRINT_DONE, Events.PRINT_CANCELLED, Events.PRINT_FAILED, ]:
            if self.timer_screen:
                self.timer_screen.cancel()
                self.timer_screen = None

            if not self.timer_seconds:
                self.timer_seconds = RepeatedTimer( 15.0, self.on_timer_seconds )

            if event in [ Events.PRINT_STARTED, ]:
                self.screen_priority_state = STATE_PRINTING
                self.printing_filename = payload['name']
                self.start_timestamp = datetime.now()
                self.update_screen_TextFileName()
                self.update_screen_TextETA()
                self.update_screen_TextPercent()
                self.update_screen_TextFIN()

                if not self.timer_seconds.is_alive():
                    self.timer_seconds.start()

                self.update_screen_priority()

            if event in [ Events.PRINT_DONE, Events.PRINT_CANCELLED, Events.PRINT_FAILED, ]:
                self.screen_priority_state = STATE_NON_PRINTING
                self.start_timestamp = None
                self.printing_eta = None
                self.printing_percent = None

                self.timer_seconds.cancel()
                self.timer_seconds = None

                if self._settings.get_boolean(["hide_page_when_idle"]):
                    self.timer_screen = ResettableTimer( 60 * self._settings.get_int(["idle_time_minutes"]), self.on_timer_screen )
                    self.timer_screen.start()

                self.update_screen_priority()

    def update_screen_TextFileName(self):
        if self.printing_filename is None:
            visible_filename = " - "
        else:
            visible_filename = self.printing_filename

        screen, screen_width, screen_height = self.ensure_screen('OctPriSCR1')
        if screen and 'TextFileName' in screen.widgets:
            self._logger.info("LCDd 'TextFileName' == '%s'" % visible_filename )
            screen.widgets['TextFileName'].set_text( visible_filename )
            screen.widgets['TextFileName'].update()

    def update_screen_priority(self):
        screen, screen_width, screen_height = self.ensure_screen('OctPriSCR1')
        if screen:
            if self.screen_priority_state == STATE_IDLE:
                if self._settings.get_boolean(["hide_page_when_idle"]):
                    self._logger.info("Switching screen priority: hidden")
                    self.lcd.screens['OctPriSCR1'].set_priority("hidden")
                else:
                    self.screen_priority_state = STATE_NON_PRINTING

            if self.screen_priority_state == STATE_NON_PRINTING:
                new_priority = self._settings.get(["priority_non_printing"])
                self._logger.info("Switching screen priority: %s" % new_priority )
                self.lcd.screens['OctPriSCR1'].set_priority( new_priority )

            if self.screen_priority_state == STATE_PRINTING:
                new_priority = self._settings.get(["priority_printing"])
                self._logger.info("Switching screen priority: %s" % new_priority )
                self.lcd.screens['OctPriSCR1'].set_priority( new_priority )

    def update_screen_TextPercent(self):
        if self.printing_percent is None:
            visible_percent = " - "
        else:
            visible_percent = "%d%%" % ( self.printing_percent )

        screen, screen_width, screen_height = self.ensure_screen('OctPriSCR1')
        if screen and 'TextPercent' in screen.widgets:
            self._logger.info("LCDd 'TextPercent' == '%s'" % visible_percent )
            screen.widgets['TextPercent'].set_text( visible_percent )
            screen.widgets['TextPercent'].set_x( screen_width - ( len( visible_percent ) - 1 ) )
            screen.widgets['TextPercent'].update()

    def update_screen_TextETA( self ):
        if self.printing_eta is None:
            visible_eta = " - "
        else:
            # limiting display to 99 hour 60 minutes
            eta_calcwith = self.printing_eta if self.printing_eta < 360000 else 360000 - 1
            eta_hours = eta_calcwith // 3600
            eta_minutes = ( eta_calcwith - ( eta_hours * 3600 ) ) // 60
            visible_eta = "%02d:%02d" % ( eta_hours, eta_minutes )

        screen, screen_width, screen_height = self.ensure_screen('OctPriSCR1')
        if screen and 'TextETA' in screen.widgets:
            self._logger.info("LCDd 'TextETA' == '%s'" % visible_eta )
            screen.widgets['TextETA'].set_text( visible_eta )
            screen.widgets['TextETA'].update()

    def update_screen_TextFIN( self ):
        if self.printing_eta is None:
            visible_fin = " - "
        else:
            FINISH_DATETIME = datetime.now() + timedelta( seconds = self.printing_eta )
            NOW_DATE = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
            FIN_DATE = FINISH_DATETIME.replace(hour=0, minute=0, second=0, microsecond=0)

            if ( (FIN_DATE-NOW_DATE).days < 1 ):
                visible_fin = "%02d:%02d" % ( FINISH_DATETIME.hour, FINISH_DATETIME.minute )
            else:
                visible_fin = "+%dd %02d:%02d" % ( (FIN_DATE-NOW_DATE).days, FINISH_DATETIME.hour, FINISH_DATETIME.minute )

        screen, screen_width, screen_height = self.ensure_screen('OctPriSCR1')
        if screen and 'TextFIN' in screen.widgets:
            self._logger.info("LCDd 'TextFIN' == '%s'" % visible_fin )
            screen.widgets['TextFIN'].set_text( visible_fin )
            screen.widgets['TextFIN'].set_x( screen_width - len( visible_fin ) )
            screen.widgets['TextFIN'].update()


    def on_print_progress(self, storage, path, progress ):
        self.printing_percent = progress

        self.update_screen_TextPercent()

    def on_timer_seconds(self):
        try:
            self.printing_eta = self._printer.get_current_data()['progress']['printTimeLeft']
        except:
            self.printing_eta = None

        self.update_screen_TextPercent()
        self.update_screen_TextETA()
        self.update_screen_TextFIN()

    def on_timer_screen(self):
        self.timer_screen = None
        self.screen_priority_state = STATE_IDLE
        self.update_screen_priority()

    def initialize_lcd(self):
        if not self._settings.get_boolean(["enabled"]):
            self.lcd = None
            self._logger.info("Connection to LCDd are disabled")
            return False

        try:
            self.lcd = Server(hostname=self._settings.get(["host"]), port=self._settings.get_int(["port"]), debug=False)
            self.lcd.start_session()
        except:
            self.lcd = None
            self._logger.exception("Unable to establish the connection to the LCDd")
            return False

        self.lcd.add_screen("OctPriSCR1")

        if self._settings.get_boolean(["title_show"]):
            first_linenum = 2
            self.lcd.screens['OctPriSCR1'].add_title_widget("TitleText", text = self._settings.get(["title_text"]) )
            self.lcd.screens['OctPriSCR1'].set_heartbeat("on")
        else:
            first_linenum = 1
            self.lcd.screens['OctPriSCR1'].set_heartbeat("off")

        self.update_screen_priority()

        self.lcd.screens['OctPriSCR1'].add_string_widget("TextPercent", text="", y= first_linenum+0, x=self.lcd.server_info['screen_width']-3 )
        self.lcd.screens['OctPriSCR1'].add_scroller_widget("TextFileName", text="", speed=5, left=1, top=first_linenum+0, right=self.lcd.server_info['screen_width']-5, bottom=first_linenum+0 )
        self.lcd.screens['OctPriSCR1'].add_string_widget("TextETA", text="", y=first_linenum+1,x=2,)
        self.lcd.screens['OctPriSCR1'].add_string_widget("TextFIN", text="", y=first_linenum+1,x=self.lcd.server_info['screen_width']-2)
        self.lcd.screens['OctPriSCR1'].add_icon_widget("IconETA", x=1, y=first_linenum+1, name="SELECTOR_AT_RIGHT" )
        self.lcd.screens['OctPriSCR1'].add_icon_widget("IconFIN", x=self.lcd.server_info['screen_width'], y=first_linenum+1, name="SELECTOR_AT_LEFT" )

        self._logger.info("LCDd connection established")

        self.update_screen_TextFileName()
        self.update_screen_TextPercent()
        self.update_screen_TextETA()
        self.update_screen_TextFIN()

        return True

    def ensure_screen(self, ref):
        if not self.lcd or not self.lcd.alive_session():
            if not self.initialize_lcd():
                return ( None, None, None )

        if self.lcd and self.lcd.alive_session() and ref in self.lcd.screens:
            return ( self.lcd.screens[ref], self.lcd.server_info['screen_width'], self.lcd.server_info['screen_height'] )

        return ( None, None, None )

__plugin_name__ = "LCDproc"

# Set the Python version your plugin is compatible with below. Recommended is Python 3 only for all new plugins.
__plugin_pythoncompat__ = ">=3.7,<4"

def __plugin_load__():
    global __plugin_implementation__
    __plugin_implementation__ = LcdprocPlugin()

    global __plugin_hooks__
    __plugin_hooks__ = {
        "octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information
    }
